File: packages/js/src/schemas/academic.ts
import { z } from "zod";

import { BaseResponseSchema } from "./common.js";

export const AcademicCourseSchema = z.object({
  codCurso: z.string(),
  ciclo: z.number(),
  desAsignatura: z.string(),
  desCalificacion: z.string(),
  codTipoActa: z.string(),
  numCreditaje: z.number(),
  valCalificacionFinal: z.number(),
  codPlan: z.string(),
  codSeccion: z.string(),
  numActa: z.string(),
  tipoAsignatura: z.string(),
  numResConv: z.unknown().nullable(),
}).passthrough();

export const AcademicHistoryResponseSchema = BaseResponseSchema.extend({
  data: z.array(z.object({
    codSemestre: z.string(),
    asignatura: z.array(AcademicCourseSchema),
  }).passthrough()),
}).passthrough();

export type AcademicHistoryResponse = z.infer<typeof AcademicHistoryResponseSchema>;

File: packages/python/src/unmsm_sum_api/schemas/academic.py
from typing import Any

from .common import BaseResponse, FlexibleModel


class AcademicCourse(FlexibleModel):
    codCurso: str
    ciclo: int
    desAsignatura: str
    desCalificacion: str
    codTipoActa: str
    numCreditaje: float
    valCalificacionFinal: float
    codPlan: str
    codSeccion: str
    numActa: str
    tipoAsignatura: str
    numResConv: Any | None


class AcademicSemester(FlexibleModel):
    codSemestre: str
    asignatura: list[AcademicCourse]


class AcademicHistoryResponse(BaseResponse):
    data: list[AcademicSemester]
